File: src/aci_cache/strategies/eager.py
"""Eager Strategy — per-write invalidation via Pub/Sub.

Every ``set()`` call immediately publishes an invalidation message so that
all other subscribed instances can delete the stale key.
"""

from __future__ import annotations

import json
import logging
import time

import redis

from .base import Strategy

logger = logging.getLogger(__name__)


class EagerStrategy(Strategy):
    """Publishes an invalidation message on every write for minimum staleness."""

    def __init__(
        self,
        pubsub_client: redis.Redis,
        channel: str,
        instance_id: str,
    ) -> None:
        self._pubsub_client = pubsub_client
        self._channel = channel
        self._instance_id = instance_id

    @property
    def name(self) -> str:
        return "eager"

    def on_write(self, key: str) -> None:
        payload = {
            "action": "invalidate",
            "keys": [key],
            "strategy": "eager",
            "timestamp": time.time(),
            "source": self._instance_id,
        }
        try:
            self._pubsub_client.publish(self._channel, json.dumps(payload))
        except redis.ConnectionError as exc:
            logger.warning("[EAGER] Publish failed (Redis unavailable): %s", exc)

    def on_activate(self) -> None:
        logger.debug("[EAGER] Strategy activated")

    def on_deactivate(self) -> None:
        logger.debug("[EAGER] Strategy deactivated")

File: node/src/strategies/eager.ts
/**
 * Eager Strategy — per-write invalidation via Pub/Sub.
 *
 * Every set() call immediately publishes an invalidation message so that
 * all other subscribed instances can delete the stale key.
 */

import type Redis from "ioredis";
import { Strategy } from "./base";
import type { InvalidationMessage } from "../types";

export class EagerStrategy implements Strategy {
  readonly name = "eager";

  private readonly pubsubClient: Redis;
  private readonly channel: string;
  private readonly instanceId: string;

  constructor(pubsubClient: Redis, channel: string, instanceId: string) {
    this.pubsubClient = pubsubClient;
    this.channel = channel;
    this.instanceId = instanceId;
  }

  async onWrite(key: string): Promise<void> {
    const payload: InvalidationMessage = {
      action: "invalidate",
      keys: [key],
      strategy: "eager",
      timestamp: Date.now() / 1000,
      source: this.instanceId,
    };
    try {
      await this.pubsubClient.publish(this.channel, JSON.stringify(payload));
    } catch {
      // Swallow connection errors — TTL safety net handles staleness
    }
  }

  onActivate(): void {
    // Nothing to do
  }

  async onDeactivate(): Promise<void> {
    // Nothing to do
  }
}
